app/assets/javascripts/provider/admin/apiconfig/services/service_discovery.js

```diff
@@ -0,0 +1,112 @@
+document.addEventListener("DOMContentLoaded", function(e) {
+
+  const formSource = document.getElementById('new_service_source');
+  const formScratch = document.querySelectorAll('form#new_service, form#create_service')[0];
+  const formDiscover = document.getElementById('service_discovery');
+
+  function clearSelectOptions(select) {
+    const options = select.getElementsByTagName('option');
+    for (var i = options.length -1; i > 0; i--) {
+      options[i].remove();
+    }
+  };
+
+  function clearNamespaces() {
+    return clearSelectOptions(formDiscover.service_namespace);
+  };
+
+  function clearServiceNames() {
+    return clearSelectOptions(formDiscover.service_name);
+  };
+
+  function showAndHideForms(showForm, hideForm) {
+    showForm.classList.remove('is-hidden');
+    hideForm.classList.add('is-hidden');
+  }
+
+  function showServiceForm(source, discoverFunc, scratchFunc) {
+    if (source === 'discover') {
+      showAndHideForms(formDiscover, formScratch);
+      return typeof discoverFunc === "function" ? discoverFunc() : undefined;
+    } else {
+      showAndHideForms(formScratch, formDiscover);
+      return typeof scratchFunc === "function" ? scratchFunc() : undefined;
+    }
+  };
+
+  function changeServiceSource() {
+    clearNamespaces();
+    clearServiceNames();
+    return showServiceForm(formSource.source.value, fetchNamespaces);
+  };
+
+  function changeClusterNamespace() {
+    clearServiceNames();
+    const selectedNamespace = formDiscover.service_namespace.value;
+    return selectedNamespace !== '' ? fetchServices(selectedNamespace) : undefined;
+  };
+
+  function addOptionToSelect(selectElem, val) {
+    var opt = document.createElement('option');
+    opt.text = val;
+    opt.value = val;
```

Comment:
Do we need to set `opt.value` if it's always equal to `opt.text`?

Reply:
Not really, we can omit that